src/Client.h

```diff
@@ -140,21 +141,48 @@ class Client : public ClientBase{
 		int serverNo = 0;
 
 		TicksTime rxTime;
-
-		ret = msg_recvfrom(ifd,
-				           g_fds_array[ifd]->recv.cur_addr + g_fds_array[ifd]->recv.cur_offset,
-				           g_fds_array[ifd]->recv.cur_size,
-				           &recvfrom_addr);
-		if (ret == RET_SOCKET_SHUTDOWN) {
-			if (g_fds_array[ifd]->sock_type == SOCK_STREAM) {
-				exit_with_log("A connection was forcibly closed by a peer",SOCKPERF_ERR_SOCKET,g_fds_array[ifd]);
+#ifdef  USING_VMA_EXTRA_API
+		if (!g_vma_api || !g_pApp->m_const_params.is_vmapoll)
+#endif
+		{
+			ret = msg_recvfrom(ifd,
+					   g_fds_array[ifd]->recv.cur_addr + g_fds_array[ifd]->recv.cur_offset,
+					   g_fds_array[ifd]->recv.cur_size,
+					   &recvfrom_addr);
+			if (ret == RET_SOCKET_SHUTDOWN) {
+				if (g_fds_array[ifd]->sock_type == SOCK_STREAM) {
+					exit_with_log("A connection was forcibly closed by a peer",SOCKPERF_ERR_SOCKET,g_fds_array[ifd]);
+				}
 			}
+			if (ret < 0) return 0;
 		}
-		if (ret < 0) return 0;
+#ifdef USING_VMA_EXTRA_API
+		int nbytes = ret;
+		vma_buff_t* tmp_vma_poll_buff = g_vma_poll_buff;
+		while(tmp_vma_poll_buff || nbytes) {
+			if (tmp_vma_poll_buff && !nbytes){
+				if (g_fds_array[ifd]->recv.cur_offset) {
+					g_fds_array[ifd]->recv.cur_addr = g_fds_array[ifd]->recv.buf;
+					memcpy(g_fds_array[ifd]->recv.cur_addr + g_fds_array[ifd]->recv.cur_offset,(uint8_t*)tmp_vma_poll_buff->payload, tmp_vma_poll_buff->len);
+				}
+				else {
+					g_fds_array[ifd]->recv.cur_addr = (uint8_t*)tmp_vma_poll_buff->payload;
+				}
+				ret = tmp_vma_poll_buff->len;
+				recvfrom_addr = g_vma_comps->src;
+				if (ret == RET_SOCKET_SHUTDOWN) {
+					if (g_fds_array[ifd]->sock_type == SOCK_STREAM) {
+						exit_with_log("A connection was forcibly closed by a peer",SOCKPERF_ERR_SOCKET,g_fds_array[ifd]);
+					}
+				}
+				if (ret < 0) return 0;
+				nbytes = ret;
+			}
 
```

Comment:
I see following ways that can be applied here:
1. current from @AlaAyesh : at least two places in source code have big block specific for vma_poll() processing as https://github.com/Mellanox/sockperf/pull/76/files#diff-100f024c25446e375ab03f92192af26dR159
2. move this code under msg_recvfrom() as was done for VMA_ZCOPY but probably it requires checking header twice (during buffer processing and outside in usual flow as https://github.com/AlaAyesh/sockperf/blob/9ac410864265ef12849beb89234d37d6b6c0c4cb/src/Client.h#L194)
3. may be we could match list processing (for vma_poll()) to iov (used for VMA_ZCOPY). In this case we could reuse VMA_ZCOPY code inside/outside.

Probably option-1 is better for performance comparing with 2,3. But using 2,3 makes code readable and slim. @AlaAyesh please correct me if I am missing something.
@rosenbaumalex @AlaAyesh @avnerbh What do you think?